stix2/sources/memory.py

```diff
@@ -0,0 +1,297 @@
+        """
+        Note: It doesnt make sense to create a MemoryStore by passing
+         in existing MemorySource and MemorySink because there could
+         be data concurrency issues. Just as easy to create new MemoryStore.
+        """
+        self.name = name
+        self.id = make_id()
+        self.data = {}
+
+        if stix_data:
+            if type(stix_data) == dict:
+                # stix objects are in a bundle
+                # verify STIX json data
+                r = validate_string(json.dumps(stix_data))
+                # make dictionary of the objects for easy lookup
+                if r.is_valid:
+                    for stix_obj in stix_data["objects"]:
+
+                        self.data[stix_obj["id"]] = stix_obj
+                else:
+                    print("Error: json data passed to MemorySink() was found to not be validated by STIX 2 Validator")
+                    print(r)
+            elif type(stix_data) == list:
+                # stix objects are in a list
+                for stix_obj in stix_data:
+                    r = validate_string(json.dumps(stix_obj))
+                    if r.is_valid:
+                        self.data[stix_obj["id"]] = stix_obj
+                    else:
+                        print("Error: STIX object %s is not valid under STIX 2 validator.") % stix_obj["id"]
+                        print(r)
+
+        self.source = MemorySource(stix_data=self.data, _store=True)
+        self.sink = MemorySink(stix_data=self.data, _store=True)
+
+    @property
+    def source(self):
+        return self.source
+
+    @property
+    def sink(self):
+        return self.sink
+
+    # memory sink API calls
+
+    def add(self, stix_data):
+        return self.sink.add(stix_data=stix_data)
+
+    def save_to_file(self, file_path):
+        return self.sink.save(file_path=file_path)
+
+    # memory source API calls
+
+    def get(self, stix_id):
+        return self.source.get(stix_id=stix_id)
+
+    def all_versions(self, stix_id):
+        return self.source.all_versions(stix_id=stix_id)
+
+    def query(self, query):
+        return self.source.query(query=query)
+
+    def load_from_file(self, file_path):
+        return self.source.load_from_file(file_path=file_path)
+
+
+class MemorySink(DataSink):
+    """
+
+    """
+    def __init__(self, stix_data=None, name="MemorySink", _store=False):
+        """
+        Args:
+
+            data (dictionary OR list): valid STIX 2.0 content in bundle or a list
+            name (string): optional name tag of the data source
+            _store (bool): if the MemorySink is a part of a DataStore, in which case
+                           "stix_data" is a direct reference to shared memory with DataSource
+
+        """
+        super(MemorySink, self).__init__(name=name)
+
+        if _store:
+            self.data = stix_data
+        else:
+            self.data = {}
+            if stix_data:
+                if type(stix_data) == dict:
```

Comment:
This won't catch the case where stix_data is an instance of one of our SDO classes. From the docstring it looks like this might be intentional, but it also won't catch the case where stix_data is an instance of our Bundle class.